### PR TITLE
feat: trigger a callback function provided by client when a token expires

### LINKED ISSUE #25: Define onTokenExpired event
**Is your feature request related to a problem? Please describe.**

Currently SDK users need to implement a token refresh process if they want to keep a user logged in. This works fine, but could be easier for SDK users when a token expired, as token expiration HTTP 401 may require some logic to be differentiated from regular HTTP 401 status code.

**Describe the solution you'd like**

**Describe alternatives you've considered**

Check for HTTP status code equals 401 and send a token refresh request.

Pseudo-code, yet to be tested:
```typescript
    public onTokenExpired: () => void;

    public tokenExpired(): void {
        if (this.onTokenExpired) {
            this.onTokenExpired();
        }
    }
```

**Additional context**

The responsibility on making the decision of refreshing a token or not doesn't change. It'll still be up to the SDK consumer to decide what to do. 
We just want to assume the logic that will check the `HTTP 401 Unauthorized` response to tell which ones are related to token expiration and notify the client app.

### PR DESCRIPTION
**Pull request by dudunog**
## Description of changes

- [x] Implement a mechanism to trigger a user-provided callback function when an access token expires. This allows the client to handle token expiration events in a custom manner, such as refreshing the token.

## GitHub issues resolved by this PR

- [x] This PR resolves #25 

## Quality Assurance

- Once the changes in this PR are merged and deployed, success criteria is: trigger a callback function provided by client when a token expires.

### CODE REVIEW

File: packages/sdks/javascript/src/services/ApiBaseService.ts [hunk 1]
```diff
@@ -120,6 +126,22 @@ export class ApiBaseService {
         ApiBaseService._accessToken = value;
     }
 
+    public static set onTokenExpired(callback: (accessToken: string) => void) {
+        ApiBaseService._onTokenExpired = callback;
+    }
+
+    protected isTokenExpired(response: AxiosResponse | undefined): boolean {
```
Comment (jfoliveira) [inline packages/sdks/javascript/src/services/ApiBaseService.ts:133]: @dudunog as all those new methods/properties might not be straightforward to all devs consuming or maintaining this API SDK, could you please add [JSDoc comments ](https://jsdoc.app/about-getting-started) to each method?

**Review by jfoliveira — CHANGES_REQUESTED** [review]

File: packages/sdks/javascript/src/services/ApiBaseService.ts (hunk 1 above)
Comment (dudunog) [reply packages/sdks/javascript/src/services/ApiBaseService.ts:133]: @JeffersonAbreu Of course, but it's only one method that the user consumes to set a callback function (onTokenExpired, inside of DevopnessApiClient), right? Are there no others?

File: packages/sdks/javascript/tests/DevopnessApiClient.spec.ts [hunk 2]
```diff
@@ -50,4 +55,20 @@ test("base URL is configurable on initialization", () => {
...
   expect(apiClient).toHaveProperty('variables')
+
+  expect(apiClient).toHaveProperty('onTokenExpired')
 })
+
+test("expired access_token should trigger callback function", async () => {
+  const mockRefreshTokenCallback = jest.fn(token => token);
+
+  const apiClient = new DevopnessApiClient();
+  apiClient.accessToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE3MTE3MjE4MTQsImV4cCI6MTcxMTYzNTQxNCwiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSIsIkdpdmVuTmFtZSI6IkpvaG5ueSIsIlN1cm5hbWUiOiJSb2NrZXQiLCJFbWFpbCI6Impyb2NrZXRAZXhhbXBsZS5jb20iLCJSb2xlIjpbIk1hbmFnZXIiLCJQcm9qZWN0IEFkbWluaXN0cmF0b3IiXX0.Id1BkqYEZUEdPyu2BaTH9ulETZiZuSH9XJVi49up1Qw';
```
Comment (jfoliveira) [inline packages/sdks/javascript/tests/DevopnessApiClient.spec.ts:66]: Perhaps add a sample token that is closer to reality? :-) 

image


```suggestion
  apiClient.accessToken = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJhcGkuZGV2b3BuZXNzLmNvbSIsImlhdCI6MTcxMjYwOTI1NiwiZXhwIjoxNzQ0MTQ1MjU2LCJhdWQiOiJ3d3cubXktcHJvZHVjdC5jb20iLCJzdWIiOiJzb21lb25lQGV4YW1wbGUuY29tIiwic2NvcGUiOlsic2VydmVyOnJlYWQiLCJzZXJ2ZXI6Z2V0X3N0YXR1cyIsImFwcGxpY2F0aW9uOmRlcGxveSIsImFwcGxpY2F0aW9uOmVkaXQiLCJjcm9uam9iOnJlYWQiLCJjcm9uam9iOmRlcGxveSIsInNzbC1jZXJ0aWZpY2F0ZTpkZXBsb3kiXX0.pEVwmAznm75UFCAsTVEqsi40TjeZDUuN86Z6QdpMyaQ;
```

**Review by jfoliveira — CHANGES_REQUESTED** [review]

File: packages/sdks/javascript/src/services/ApiBaseService.ts (hunk 1 above)
Comment (jfoliveira) [reply packages/sdks/javascript/src/services/ApiBaseService.ts:133]: That's fine! Adding JSDoc comments to the public methods exposed to the SDK user should be enough.

File: packages/sdks/javascript/tests/DevopnessApiClient.spec.ts (hunk 2 above)
Comment (dudunog) [reply packages/sdks/javascript/tests/DevopnessApiClient.spec.ts]: Improved!

File: packages/sdks/javascript/src/DevopnessApiClient.ts [hunk 3]
```diff
@@ -118,4 +121,17 @@ export class DevopnessApiClient {
...
+   *
+   * @summary Sets the callback function to be executed when the access token expires.
+   * @static
+   * @param {function(string)} callback - The callback function to be executed.
+   * This function should accept a single string parameter representing the expired access token.
+   */
+  public set onTokenExpired(callback: (accessToken: string) => void) {
+    if (callback) {
+      ApiBaseService.onTokenExpired = callback;
+    }
```
Comment (thlmenezes) [inline packages/sdks/javascript/src/DevopnessApiClient.ts:135]: @dudunog why use a conditional block here? `callback` is non-optional by the types on the function declaration above.

```suggestion
    ApiBaseService.onTokenExpired = callback;
```

Comment (dudunog) [reply packages/sdks/javascript/src/DevopnessApiClient.ts]: Ops, this makes sense, we don't need a conditional block there.

**Comment by dudunog** [comment]
I have read the CLA document and I hereby sign the CLA.

**Comment by dudunog** [comment]
I have read the CLA document and I hereby sign the CLA

**Review by thlmenezes — APPROVED** [review]

**Review by jfoliveira — APPROVED** [review]

**Pull request merged**